olive/common/ort_inference.py

```diff
@@ -51,26 +75,299 @@ def get_ort_inference_session(
         for key, value in extra_session_config.items():
             sess_options.add_session_config_entry(key, value)
 
-    if isinstance(execution_provider, str):
-        execution_provider = [execution_provider]
-    else:
-        # execution providers should be list[str]
-        assert isinstance(execution_provider, list)
-        assert all(isinstance(ep, str) for ep in execution_provider)
-
-    for idx, ep in enumerate(execution_provider):
-        if ep == "QNNExecutionProvider":
+    # execution providers and provider options
+    providers, provider_options = check_and_normalize_provider_args(
+        inference_settings.get("execution_provider"),
+        inference_settings.get("provider_options"),
+        ort.get_available_providers(),
+    )
+    for idx, provider in enumerate(providers):
+        if provider in ["CUDAExecutionProvider", "DmlExecutionProvider"] and device_id is not None:
+            provider_options[idx]["device_id"] = str(device_id)
+        elif provider == "QNNExecutionProvider":
             # add backend_path for QNNExecutionProvider
-            execution_provider[idx] = ("QNNExecutionProvider", {"backend_path": "QnnHtp.dll"})
-            break
+            provider_options[idx]["backend_path"] = "QnnHtp.dll"
```

Comment:
Not have to update in this PR.
We may need expose the backend_path to users? Also need to handle different suffix in different platform. Linux -> .so, Windows -> .dll.

Could you help add the TODO to tell this?

Reply:
I checked the qnn ep docs and since there is only windows arm package currently, their examples also only use htp backend (the other option is cpu backend but that's not interesting). 
The expectation for qnn ep is that they already have the qnn sdk path added to `PATH` so that it can find the backend library using the name.

Reply:
Ic, sounds like we can focus on windows by now as qnn ep is only available in that.